presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryPagesStore.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.memory;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.Block;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+
+@ThreadSafe
+public class MemoryPagesStore
+{
+    private final Map<Long, List<Page>> pages = new HashMap<>();
+
+    public synchronized void add(Long tableId, Page page)
+    {
+        if (!pages.containsKey(tableId)) {
+            pages.put(tableId, new ArrayList<>());
+        }
+
+        List<Page> tablePages = pages.get(tableId);
+        tablePages.add(page);
+    }
+
+    public synchronized List<Page> getPages(Long tableId, int partNumber, int totalParts, List<Integer> columnIndexes)
+    {
+        checkState(pages.containsKey(tableId));
+
+        List<Page> tablePages = pages.get(tableId);
+        ImmutableList.Builder<Page> partitionedPages = ImmutableList.builder();
+
+        for (int i = partNumber; i < tablePages.size(); i += totalParts) {
+            partitionedPages.add(reorderPage(tablePages.get(i), columnIndexes));
+        }
+
+        return partitionedPages.build();
+    }
+
+    public synchronized boolean contains(Long tableId)
+    {
+        return pages.containsKey(tableId);
+    }
+
+    public synchronized void cleanUp(Set<Long> activeTableIds)
+    {
+        // We have to remember that there might be some race conditions when there are two tables created at once.
+        // That can lead to a situation when MemoryPagesStore already knows about a newer second table on some worker
+        // but cleanUp is triggered by insert from older first table, which MemoryTableHandle was created before
+        // second table creation. Thus activeTableIds can have missing latest ids and we can only clean up tables
+        // that have smaller value then max(activeTableIds).
```

Comment:
Case1:

1. insertion into table 100 starts. On coordinator `activeTableIds = {100}`
2. few pages are written on some workers for table `100`. Every worker that has written some pages knows that `activeTableIds = {100}`
3. insertion/creation on table 101 starts. On coordinator `activeTableIds = {100, 101}`
4. if any worker gets some pages from 101 before 100, nothing dramatic happens. No pages are wiped out, because:
  - any `cleanUp` for call with `activeTableIds = {100}` will never delete pages from tables > 100
  - any `cleanUp` for `activeTableIds = {100, 101}` will not drop any pages from 100 nor 101, because they are "active".


After re-reading my comment I know what has mislead you. Sentence: `and we can only clean up tables that have smaller value then max(activeTableIds)`. "Can" doesn't mean it will be delete. Condition for deletion is that table must have:
- `tableId < max(activeTableIds)`
- and table id must be absent in `activeTableIds`

I have modified commit to make that clear.